agent_app_sample_code/utils/get_recursive_character_text_splitter.py

```diff
@@ -0,0 +1,180 @@
+# MAGIC
+# MAGIC **Returns:** A callable that takes a document (`str`) and produces a list of chunks (`list[str]`).
+
+# COMMAND ----------
+
+from langchain_text_splitters import RecursiveCharacterTextSplitter
+from transformers import AutoTokenizer
+import tiktoken
+from typing import Callable
+import os
+from databricks.sdk import WorkspaceClient
+
+
+# This is necessary because spark sometimes will give a warning about the cache in `/.cache` being read-only.
+HF_CACHE_DIR = "/tmp/hf_cache/"
+
+
+EMBEDDING_MODELS = {
+    "gte-large-en-v1.5": {
+        "tokenizer": lambda: AutoTokenizer.from_pretrained(
+            "Alibaba-NLP/gte-large-en-v1.5", cache_dir=HF_CACHE_DIR
+        ),
+        "context_window": 8192,
+        "type": "SENTENCE_TRANSFORMER",
+    },
+    "bge-large-en-v1.5": { # FMAPI Per-Token name
+        "tokenizer": lambda: AutoTokenizer.from_pretrained(
+            "BAAI/bge-large-en-v1.5", cache_dir=HF_CACHE_DIR
+        ),
+        "context_window": 512,
+        "type": "SENTENCE_TRANSFORMER",
+    },
+    "bge_large_en_v1_5": { # FMAPI PT name
+        "tokenizer": lambda: AutoTokenizer.from_pretrained(
+            "BAAI/bge-large-en-v1.5", cache_dir=HF_CACHE_DIR
+        ),
+        "context_window": 512,
+        "type": "SENTENCE_TRANSFORMER",
+    },
+    "text-embedding-ada-002": {
+        "context_window": 8192,
+        "tokenizer": lambda: tiktoken.encoding_for_model("text-embedding-ada-002"),
+        "type": "OPENAI",
+    },
+    "text-embedding-3-small": {
+        "context_window": 8192,
+        "tokenizer": lambda: tiktoken.encoding_for_model("text-embedding-3-small"),
+        "type": "OPENAI",
+    },
+    "text-embedding-3-large": {
+        "context_window": 8192,
+        "tokenizer": lambda: tiktoken.encoding_for_model("text-embedding-3-large"),
+        "type": "OPENAI",
+    },
+}
+
+
+def detect_fmapi_embedding_model_type(model_serving_endpoint):
+    """
+    Try to detect the endpoint type and get the embedding config
+    returns (None, None) if not found
+    returns (endpoint_type, embedding_config)
+    """
+    w = WorkspaceClient()
+
+    endpoint_type = None
+    try:
+        llm_endpoint = w.serving_endpoints.get(name=model_serving_endpoint)
+        # external model name
+        endpoint_type = llm_endpoint.config.served_entities[0].external_model.name
+    except Exception as e:
+        try:
+            # FMAPI pay per token
+            endpoint_type = llm_endpoint.config.served_entities[0].foundation_model.name
+        except Exception as e:
+            try: 
+                # FMAPI provisioned throughput
+                endpoint_type = llm_endpoint.config.served_entities[0].name
+                if endpoint_type[-2:][0] == '-':
+                    # remove the version number if present e.g., `bge_large_en_v1_5-2` --> `bge_large_en_v1_5`
+                    endpoint_type = endpoint_type[:-2]
```

Comment:
Hm, is this the recommended way to do this?

Reply:
the served entity on databricks has the version so this is needed to match the type. i replaced this with regex to make it less hacky tho.